mesos-rxjava-core/src/main/java/org/apache/mesos/rx/java/MesosSchedulerClient.java

```diff
@@ -185,4 +169,123 @@ public Subscription sink(@NotNull final Observable<SinkOperation<Send>> spout) {
         };
     }
 
+    /**
+     * Turns an Observable into and AwaitableSubscription
+     */
+    private static final class ObservableAwaitableSubscription implements AwaitableSubscription {
+        @NotNull
+        private final Observable<Optional<Throwable>> observable;
+        @NotNull
+        private final Subscription subscription;
+
+        public ObservableAwaitableSubscription(
+            @NotNull final Observable<Optional<Throwable>> observable,
+            @NotNull final Subscription subscription
+        ) {
+            this.observable = observable;
+            this.subscription = subscription;
+        }
+
+        @Override
+        public void await() throws Throwable {
+            final Optional<Throwable> last = observable.toBlocking().last();
+            subscription.unsubscribe();
+            if (last.isPresent()) {
+                throw last.get();
+            }
+        }
+
+        @Override
+        public void unsubscribe() {
+            subscription.unsubscribe();
+        }
+
+        @Override
+        public boolean isUnsubscribed() {
+            return subscription.isUnsubscribed();
+        }
+    }
+
+    /**
+     * Decorator for {@link Subscriber} that will keep track of "complete"ness and when complete will put an
+     * item onto a blocking queue. {@link Callable#call()} will then block until there is an item on the queue.
+     * If the item in the queue contains and exception, that exception will be throw back up to the user.
+     */
+    private static final class SubscriberDecorator<T> extends Subscriber<T> implements Callable<Optional<Throwable>> {
+        @NotNull
+        private final Subscriber<T> delegate;
+
+        @NotNull
+        private final BlockingQueue<Optional<Throwable>> queue;
+
+        public SubscriberDecorator(@NotNull final Subscriber<T> delegate) {
+            this.delegate = delegate;
+            queue = new LinkedBlockingDeque<>();
+        }
+
+        @Override
+        public Optional<Throwable> call() throws Exception {
+            return queue.take();
+        }
+
+        @Override
+        public void onNext(final T t) {
+            try {
+                delegate.onNext(t);
+            } catch (Exception e) {
+                onError(e);
+            }
+        }
+
+        @Override
+        public void onError(final Throwable e) {
+            Exceptions.throwIfFatal(e);
+            try {
+                delegate.onError(e);
+                queue.offer(Optional.of(e));
+            } catch (Exception e1) {
+                queue.offer(Optional.of(e1));
+            }
+        }
+
+        @Override
+        public void onCompleted() {
+            try {
+                delegate.onCompleted();
+            } catch (Exception e) {
+                queue.offer(Optional.of(e));
+            }
+            queue.offer(Optional.empty());
```

Comment:
This should probably be moved into the `try` block, otherwise it will execute even if an exception has already been added to the queue.

I assume the queue is only supposed to contain between zero and one elements? We could give it a capacity restriction and add elements to it using `add()` instead of `offer()` so that an exception is thrown if we try to add more than one element.

Add a test case for this also?